2000-11-06  Jakub Jelinek  <jakub@redhat.com>

	* decl.c (pop_binding_level): Do nothing if popping from
	global_binding_levels and the input had errors.
	(finish_function): If input had errors, pop enough binding
	levels to get at the function parameters binding level.

--- gcc/cp/decl.c.jj	Mon Nov  6 09:38:25 2000
+++ gcc/cp/decl.c	Mon Nov  6 12:09:45 2000
@@ -613,7 +613,11 @@ pop_binding_level ()
     {
       /* Cannot pop a level, if there are none left to pop.  */
       if (current_binding_level == global_binding_level)
-	my_friendly_abort (123);
+	{
+	  if (errorcount)
+	    return;
+	  my_friendly_abort (123);
+	}
     }
   /* Pop the current level, and free the structure for reuse.  */
 #if defined(DEBUG_CP_BINDING_LEVELS)
@@ -14388,7 +14392,16 @@ finish_function (flags)
   if (doing_semantic_analysis_p ())
     {
       if (current_binding_level->parm_flag != 1)
-	my_friendly_abort (122);
+	{
+	  if (errorcount)
+	    {
+	      /* This can happen on bad input.  */
+	      while (current_binding_level->parm_flag == 0)
+		pop_binding_level ();
+	    }
+	  if (current_binding_level->parm_flag != 1)
+	    my_friendly_abort (122);
+	}
       poplevel (1, 0, 1);
     }
 
